back/boxtribute_server/models/crud.py

```diff
@@ -453,69 +475,204 @@ def _update_shipment_with_prepared_boxes(*, shipment, box_label_identifiers, use
+    *, shipment, user_id, shipment_detail_update_inputs
+):
+    """Check in all given boxes by updating shipment details (target product and
+    location). Transition the corresponding box's state to 'Received'.
+    If boxes are requested to be checked-in with a location or a product that is not
+    registered in the target base, they are silently discarded.
+    """
+    # Transform input list into dict for easy look-up
+    update_inputs = {
+        int(i["id"]): {
+            "target_product_id": i["target_product_id"],
+            "target_location_id": i["target_location_id"],
+        }
+        for i in shipment_detail_update_inputs or []
+    }
+
+    details = []
+    detail_ids = tuple(update_inputs)
+    for detail in (
+        ShipmentDetail.select(ShipmentDetail, Shipment)
+        .join(Shipment)
+        .where(
+            (ShipmentDetail.shipment == shipment.id) & (ShipmentDetail.id << detail_ids)
+        )
+    ):
+        update_input = update_inputs[detail.id]
+        target_product_id = update_input["target_product_id"]
+        target_location_id = update_input["target_location_id"]
+
+        if not _validate_base_as_part_of_shipment(
+            target_location_id, detail=detail, model=Location
+        ) or not _validate_base_as_part_of_shipment(
+            target_product_id, detail=detail, model=Product
+        ):
+            continue
+
+        detail.target_product = target_product_id
+        detail.target_location = target_location_id
+        detail.box.state = BoxState.Received
+        details.append(detail)
+
+    if details:
+        Box.bulk_update([d.box for d in details], [Box.state])
+        ShipmentDetail.bulk_update(
+            details, [ShipmentDetail.target_product, ShipmentDetail.target_location]
+        )
+
+
+def _complete_shipment_if_applicable(*, shipment, user_id):
+    """If all boxes of the shipment are marked as Received or Lost, transition the
+    shipment state to 'Completed', soft-delete the corresponding shipment details,
+    assign target product and location to boxes, and transition received boxes to
+    'InStock'.
+    """
+    details = (
+        ShipmentDetail.select(ShipmentDetail, Box)
+        .join(Box)
+        .where(ShipmentDetail.shipment == shipment.id)
+    )
+    if all(d.box.state_id in [BoxState.Received, BoxState.Lost] for d in details):
+        now = utcnow()
+        shipment.state = ShipmentState.Completed
+        shipment.completed_by = user_id
+        shipment.completed_on = now
+        shipment.save()
+
+        received_boxes = []
+        for detail in details:
+            detail.deleted_by = user_id
+            detail.deleted_on = now
+            detail.box.product = detail.target_product
+            detail.box.location = detail.target_location
+            if detail.box.state_id == BoxState.Received:
+                detail.box.state = BoxState.InStock
+                received_boxes.append(detail.box)
+
```

Comment:
@pylipp Would it make sense to wrap these updates in transactions, since it seems to me that they should be atomic? What do you think?

Reply:
The function `_complete_shipment_if_applicable` is called from within an `atomic` context manager (lines 641-664). I think that should be fine then! Very glad though that you take some scrutiny for the review :D !